### PR TITLE
Separate pre plan sanity check from upgrade plan

### PR DESCRIPTION
**Pull request by chanchiwai-ray**
Run pre plan sanity check before plan generation or upgrade step, and stop the plan generation or upgrade if the pre plan sanity check failed.

### CODE REVIEW

**Comment by chanchiwai-ray** [comment]
Need to improve the UI later

**Comment by chanchiwai-ray** [comment]
Example output:  
```
$ cou plan 
Full execution log: '/home/raychan/.local/share/cou/log/cou-20240913173938.log'  
Connected to 'ceph' ✔
Analyzing cloud... ✔
Verifying cloud... ✔
2024-09-13 17:39:39 [ERROR] 'noout' should be set for ceph cluster
2024-09-13 17:39:39 [ERROR] Units in blocked: ceph-osd/1, ceph-osd/2, ceph-osd/0
2024-09-13 17:39:39 [ERROR] Running upgrades will not be possible until problems indicated in the errors are resolved

$ cou plan 
Full execution log: '/home/raychan/.local/share/cou/log/cou-20240913174023.log'  
Connected to 'ceph' ✔
Analyzing cloud... ✔
Verifying cloud... ✔
Generating upgrade plan... ✔
Upgrade cloud from 'yoga' to 'zed'
        Back up MySQL databases
        Archive old database data on nova-cloud-controller
        Control Plane principal(s) upgrade plan
                Upgrade plan for 'ceph-mon' to 'zed'
                        Upgrade software packages of 'ceph-mon' from the current APT repositories
                                Ψ Upgrade software packages on unit 'ceph-mon/0'
                        Ensure that the 'require-osd-release' option matches the 'ceph-osd' version
                        Change charm config of 'ceph-mon' 'source' to 'cloud:jammy-zed'
                        Wait for up to 2400s for model 'ceph' to reach the idle state
                        Verify that the workload of 'ceph-mon' has been upgraded on units: ceph-mon/0
        Remaining Data Plane principal(s) upgrade plan
                Upgrade plan for 'ceph-osd' to 'zed'
                        Verify that all 'nova-compute' units has been upgraded
                        Upgrade software packages of 'ceph-osd' from the current APT repositories
                                Ψ Upgrade software packages on unit 'ceph-osd/0'
                                Ψ Upgrade software packages on unit 'ceph-osd/1'
                                Ψ Upgrade software packages on unit 'ceph-osd/2'
                        Change charm config of 'ceph-osd' 'source' to 'cloud:jammy-zed'
                        Wait for up to 300s for app 'ceph-osd' to reach the idle state
                        Verify that the workload of 'ceph-osd' has been upgraded on units: ceph-osd/0, ceph-osd/1, ceph-osd/2
        Ensure ceph-mon's 'require-osd-release' option matches the 'ceph-osd' version

2024-09-13 17:40:24 [WARNING] Setting noout for ceph cluster during upgrade is optional but recommended. You can use `--set-noout` to add this optional step. For more information, Please refer to set-noout action: https://charmhub.io/ceph-mon/actions#set-noout

```

File: cou/steps/ceph.py [hunk 1]
```diff
@@ -144,18 +144,16 @@ async def osd_noout(model: Model, enable: bool) -> None:
     try:
         ceph_mon_apps = await _get_applications(model)
     except ApplicationNotFound as e:
-        logger.warning("%s", str(e))
-        logger.warning("Skip changing 'noout', because there's no ceph-mon applications.")
+        logger.debug("%s", str(e))
+        logger.debug("Skip changing 'noout', because there's no ceph-mon applications.")
```
Comment (samuelallan72) [inline cou/steps/ceph.py:148]: Can we consider changing log levels in a separate PR?  I know that we want to improve the output, but this seems like something we could do consistently across the application in a separate change. :)

**Review by samuelallan72 — CHANGES_REQUESTED** [review]
Looks like a good approach to me.  I agree though that we can improve the UI later, so I left a comment about removing the changes to log levels. :)

File: cou/steps/ceph.py (hunk 1 above)
Comment (chanchiwai-ray) [reply cou/steps/ceph.py]: done

**Review by samuelallan72 — DISMISSED** [review]
Thanks, nice change :+1:

File: cou/cli.py [hunk 2]
```diff
@@ -108,78 +113,70 @@ async def continue_upgrade() -> bool:
...
 
-    return upgrade_plan
-
+    if errors := PlanStatus.error_messages:
+        for error in errors:
+            logger.error(error)
+        raise CloudVerificationError(
+            "Running upgrades will not be possible until problems indicated in the errors "
+            "are resolved"
+        )
```
Comment (gabrielcocenza) [inline cou/cli.py:158]: I think this is a kind of regression. ManSol complained about not being able to generate and see upgrade plan if we detect some issue.

I think we should block the upgrade, but not the plan generation.

Comment (chanchiwai-ray) [reply cou/cli.py:158]: I think that's a fair request

**Comment by chanchiwai-ray** [comment]
rebased

File: cou/steps/vault.py [hunk 3]
```diff
@@ -39,5 +39,5 @@ async def verify_vault_is_unsealed(model: Model) -> None:
                     "https://charmhub.io/vault to unseal the vault manually before upgrade"
                 )
     except ApplicationNotFound:
-        logger.warning("Application vault not found, skip")
+        logger.debug("Application vault not found, skip")
```
Comment (samuelallan72) [inline cou/steps/vault.py:42]: I think this is unrelated?

File: cou/steps/plan.py [hunk 4]
```diff
@@ -265,11 +269,67 @@ def _verify_data_plane_ready_to_upgrade(args: CLIargs, analysis_result: Analysis
...
+    :type analysis_result: Analysis
+    :raises DataPlaneCannotUpgrade: When data-plane is not ready to upgrade.
+    """
+    try:
+        await ceph.assert_osd_noout_state(
+            analysis_result.model, analysis_result.apps_control_plane, state=False
+        )
+    except ApplicationError:
+        PlanStatus.add_message(
+            "'noout' should be set for ceph cluster",
```
Comment (samuelallan72) [inline cou/steps/plan.py:309]: This message contradicts the function definition - should `noout` be set or unset?

File: cou/steps/plan.py [hunk 5]
```diff
@@ -265,11 +269,67 @@ def _verify_data_plane_ready_to_upgrade(args: CLIargs, analysis_result: Analysis
...
+        )
+
+
+async def _verify_model_idle(analysis_result: Analysis) -> None:
+    """Verify the model is active idle.
+
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :raises DataPlaneCannotUpgrade: When data-plane is not ready to upgrade.
+    """
```
Comment (samuelallan72) [inline cou/steps/plan.py:320]: The docstrings to these new functions could be improved.  What does it mean to verify - the function returns None?  It states that it raises DataPlaneCannotUpgrade, but all exceptions are caught.  Why does it take an Analysis?

File: cou/steps/plan.py [hunk 6]
```diff
@@ -224,9 +226,10 @@ def _verify_supported_series(analysis_result: Analysis) -> None:
     supporting_lts_series = ", ".join(LTS_TO_OS_RELEASE)
     current_series = analysis_result.current_cloud_series
     if current_series not in LTS_TO_OS_RELEASE:
-        raise OutOfSupportRange(
+        PlanStatus.add_message(
             f"Cloud series '{current_series}' is not a Ubuntu LTS series supported by COU. "
-            f"The supporting series are: {supporting_lts_series}"
+            f"The supporting series are: {supporting_lts_series}",
```
Comment (samuelallan72) [inline cou/steps/plan.py:231]:
```suggestion
            f"The supported series are: {supporting_lts_series}",
```

File: cou/steps/plan.py [hunk 7]
```diff
@@ -129,8 +145,8 @@ async def generate_plan(analysis_result: Analysis, args: CLIargs) -> UpgradePlan
     return plan
 
 
-async def post_upgrade_sanity_checks(analysis_result: Analysis) -> None:
-    """Run post upgrade sanity checks.
+async def post_upgrade_sanity_check(analysis_result: Analysis) -> None:
+    """Run post upgrade sanity check.
```
Comment (samuelallan72) [inline cou/steps/plan.py:149]: This change seems unrelated?

File: cou/steps/plan.py [hunk 8]
```diff
@@ -95,6 +95,23 @@ def add_message(cls, message: str, message_type: MessageType = MessageType.WARNI
                 cls.warning_messages.append(message)
 
 
+async def verify_cloud(analysis_result: Analysis, args: CLIargs) -> None:
+    """Verify the cloud is ready for upgrade.
+
```
Comment (samuelallan72) [inline cou/steps/plan.py:100]: The docstring could explain what it means to verify here.  It's not obvious, because the function returns None and doesn't mention raising exceptions.

File: cou/cli.py [hunk 9]
```diff
@@ -247,6 +242,9 @@ def entrypoint() -> None:
             "$ COU_TIMEOUT=60 cou plan"
         )
         sys.exit(1)
+    except CloudVerificationError as exc:
+        logger.error(exc)
+        sys.exit(1)
```
Comment (samuelallan72) [inline cou/cli.py:247]: Can we reuse the COUException handler below, to avoid adding more code?

File: cou/cli.py [hunk 10]
```diff
@@ -217,23 +214,21 @@ async def _run_command(args: CLIargs) -> None:
     :param args: CLI arguments
     :type args: CLIargs
     """
-    match args.command:
-        case "plan":
-            await get_upgrade_plan(args)
-        case "upgrade":
-            await run_upgrade(args)
+    model = await get_model(args)
+    cloud_upgrade_plan = await analyze_and_generate_plan(model, args)
```
Comment (samuelallan72) [inline cou/cli.py:218]: We could add a comment here about why `"plan"` isn't explicitly handled here any more.

File: cou/cli.py [hunk 11]
```diff
@@ -108,78 +113,70 @@ async def continue_upgrade() -> bool:
...
-            "are resolved."
+            "are resolved"
         )
-        return
 
-    print(
-        "Please note that the actual upgrade steps could be different if the cloud state "
-        "changes because the plan will be re-calculated at upgrade time."
-    )
+    return upgrade_plan
```
Comment (samuelallan72) [inline cou/cli.py:169]: I think we still need this message, because it could still be different between running `cou plan` and `cou upgrade` - it doesn't reuse the plan from the previous invocation.

**Review by samuelallan72 — CHANGES_REQUESTED** [review]

**Review by jneo8 — CHANGES_REQUESTED** [review]

File: cou/cli.py [hunk 12]
```diff
@@ -217,23 +214,21 @@ async def _run_command(args: CLIargs) -> None:
     :param args: CLI arguments
     :type args: CLIargs
     """
-    match args.command:
-        case "plan":
-            await get_upgrade_plan(args)
-        case "upgrade":
-            await run_upgrade(args)
+    model = await get_model(args)
```
[the remark this answers is not in the document]
Comment (chanchiwai-ray) [reply cou/cli.py:217]: can we not increase the scope of this PR?

File: cou/cli.py (hunk 2 above)
Comment (chanchiwai-ray) [reply cou/cli.py]: Done

File: cou/cli.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jneo8) [reply cou/cli.py:217]: Creating model connection here break the module dependency.

[the remark this answers is not in the document]
Comment (chanchiwai-ray) [reply cou/cli.py:217]: Yes, we can rollback to existing logic. What I'm saying it only the `verify` subcommand

File: cou/cli.py (hunk 9 above)
Comment (chanchiwai-ray) [reply cou/cli.py:247]: Yes, we can refactor the UI later. `COUException` will also print out `"See the known issues at https://canonical-charmed-openstack-upgrader.readthedocs-....` which is not true for our case: we can fix the issue; it's no a known issue

File: cou/steps/plan.py (hunk 8 above)
Comment (chanchiwai-ray) [reply cou/steps/plan.py:100]: It can be too verbose here to list down what we check

File: cou/cli.py (hunk 10 above)
Comment (chanchiwai-ray) [reply cou/cli.py:228]: removed

File: cou/cli.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (chanchiwai-ray) [reply cou/cli.py:227]: FIxed

File: cou/steps/plan.py (hunk 8 above)
Comment (samuelallan72) [reply cou/steps/plan.py:100]: I mean something along the lines of "run a sequence of verification functions which populate the PlanStatus singleton with messages when errors are found".

File: cou/steps/plan.py (hunk 7 above)
Comment (chanchiwai-ray) [reply cou/steps/plan.py]: Reverted

File: cou/cli.py (hunk 11 above)
Comment (chanchiwai-ray) [reply cou/cli.py]: reverted

File: cou/cli.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jneo8) [reply cou/cli.py:227]: Thanks

**Review by jneo8 — APPROVED** [review]

**Review by samuelallan72 — APPROVED** [review]
Thanks, looks good :)
